packages/ai-jsx/src/core/conversation.tsx

```diff
@@ -0,0 +1,295 @@
+ * invocation includes the messages from the `children` prop.
+ *
+ * @example
+ * ```tsx
+ *    <Converse reply={(messages) => messages[0].type === "user" ? <AssistantMessage>Hello there!</AssistantMessage> : null}>
+ *      <UserMessage>Hello!</UserMessage>
+ *    </ChatCompletion>
+ *
+ *    ==> 'Hello there!'
+ * ```
+ *
+ */
+export async function* Converse(
+  {
+    reply,
+    children,
+  }: {
+    reply: (messages: ConversationMessage[], fullConversation: ConversationMessage[]) => AI.Renderable;
+    children: AI.Node;
+  },
+  { render, memo }: AI.ComponentContext
+): AI.RenderableStream {
+  yield AI.AppendOnlyStream;
+
+  let fullConversation = [] as ConversationMessage[];
+  let next = memo(children);
+  while (true) {
+    const messages = await renderToConversation(next, render);
+    if (messages.length === 0) {
+      break;
+    }
+
+    fullConversation = fullConversation.concat(messages);
+    next = memo(reply(messages, fullConversation));
+    yield next;
+  }
+
+  return AI.AppendOnlyStream;
+}
+
+/**
+ * Allows the presentation of conversational components ({@link UserMessage} et al) to be altered.
+ *
+ * Also accepts an `onComplete` prop, which will be invoked once per render with the entire conversation.
+ *
+ * @example
+ * ```tsx
+ *     <ShowConversation present={(msg) => msg.type === "assistant" && <>Assistant: {msg.element}</>}>
+ *         <UserMessage>This is not visible.</UserMessage>
+ *         <Assistant>This is visible!</UserMessage>
+ *     </ShowConversation>
+ *
+ *     ==> 'Assistant: This is visible!'
+ * ```
+ */
+export async function* ShowConversation(
+  {
+    children: children,
+    present,
+    onComplete,
+  }: {
+    children: AI.Node;
+    present?: (message: ConversationMessage) => AI.Node;
+    onComplete?: (conversation: ConversationMessage[], render: AI.RenderContext['render']) => Promise<void> | void;
+  },
+  { render, isAppendOnlyRender, memo }: AI.ComponentContext
+): AI.RenderableStream {
+  // If we're in an append-only render, do the transformation in an append-only manner so as not to block.
+  if (isAppendOnlyRender) {
+    yield AI.AppendOnlyStream;
+  }
+
+  let lastFrame = [] as AI.PartiallyRendered[];
+  let conversationMessages = [] as ConversationMessage[];
+
+  function handleFrame(frame: AI.PartiallyRendered[]): AI.Node {
+    if (isAppendOnlyRender) {
+      const delta = toConversationMessages(frame.slice(lastFrame.length));
+      conversationMessages = conversationMessages.concat(delta);
+      lastFrame = frame;
+      return delta.map(present ?? ((m) => m.element));
+    }
+
+    conversationMessages = toConversationMessages(frame);
+    return toConversationMessages(frame).map(present ?? ((m) => m.element));
+  }
+
+  // Memoize before rendering so that the all the conversational components get memoized as well.
+  const finalFrame = yield* render(memo(children), {
+    map: handleFrame,
+    stop: isConversationalComponent,
+    appendOnly: isAppendOnlyRender,
```

Comment:
Oh shoot, does that mean from now on we have to manually pass `appendOnly` to render? If not, then we're altering the expected behaviour which might not be okay (and previous code wasn't built with this in mind). Is there a way to pass it anyway if not provided?

Reply:
Not at all -- this component is just trying to optimize for its behavior for either scenario. If it's being rendered in append-only then it presents itself as append-only, but if not, it doesn't (e.g. allow `<FunctionCall partial>` values to surface).

Reply:
If a component presents _itself_ as append-only by yielding `AI.AppendOnlyStream` then it must only `yield` values _to be appended_.

And if a component presents itself as append-only but wants to yield intermediate values from an explicit `render` then it can only maintain the invariant by also requesting that the explicit `render` _also_ be append-only.

So, because this component conditionally presents itself as append-only, it must also conditionally make the child render append-only. (It's also currently the only component that does this wacky conditional append-only. If we find ourselves doing this in several places we should almost certainly change the abstraction.)